File: core-python/src/rappel_core/runner/value_visitor.py
"""Shared ValueExpr visitors for traversal, resolution, and evaluation."""

from __future__ import annotations

from typing import (
    TYPE_CHECKING,
    Any,
    Callable,
    Generic,
    Mapping,
    Optional,
    Sequence,
    TypeAlias,
    TypeVar,
)
from uuid import UUID

from ..dag import assert_never

if TYPE_CHECKING:  # pragma: no cover - type checkers only
    from .state import (  # noqa: F401
        ActionResultValue,
        BinaryOpValue,
        DictValue,
        DotValue,
        FunctionCallValue,
        IndexValue,
        ListValue,
        LiteralValue,
        SpreadValue,
        UnaryOpValue,
        VariableValue,
    )

ValueExpr: TypeAlias = (
    "LiteralValue | VariableValue | ActionResultValue | BinaryOpValue | UnaryOpValue | ListValue | "
    "DictValue | IndexValue | DotValue | FunctionCallValue | SpreadValue"
)

TExpr = TypeVar("TExpr")

# Cache module-level references to avoid repeated imports in hot paths.
# These are set lazily on first use to avoid circular import issues.
_STATE_CLASSES: dict[str, type] | None = None


def _get_state_classes() -> dict[str, type]:
    """Get cached state type classes, importing lazily on first call."""
    global _STATE_CLASSES
    if _STATE_CLASSES is None:
        from .state import (
            ActionCallSpec,
            ActionResultValue,
            BinaryOpValue,
            DictEntryValue,
            DictValue,
            DotValue,
            FunctionCallValue,
            IndexValue,
            ListValue,
            LiteralValue,
            SpreadValue,
            UnaryOpValue,
            VariableValue,
        )

        _STATE_CLASSES = {
            "LiteralValue": LiteralValue,
            "VariableValue": VariableValue,
            "ActionResultValue": ActionResultValue,
            "BinaryOpValue": BinaryOpValue,
            "UnaryOpValue": UnaryOpValue,
            "ListValue": ListValue,
            "DictValue": DictValue,
            "DictEntryValue": DictEntryValue,
            "IndexValue": IndexValue,
            "DotValue": DotValue,
            "FunctionCallValue": FunctionCallValue,
            "SpreadValue": SpreadValue,
            "ActionCallSpec": ActionCallSpec,
        }
    return _STATE_CLASSES


class ValueExprVisitor(Generic[TExpr]):
    """Visit ValueExpr nodes with explicit handlers.

    Example:
    - A visitor can count nodes in `a + 1` by visiting a BinaryOpValue and its children.
    """

    def visit(self, expr: "ValueExpr") -> TExpr:
        # Use type name for fast dispatch - avoids isinstance overhead
        type_name = type(expr).__name__
        if type_name == "LiteralValue":
            return self.visit_literal(expr)
        if type_name == "VariableValue":
            return self.visit_variable(expr)
        if type_name == "ActionResultValue":
            return self.visit_action_result(expr)
        if type_name == "BinaryOpValue":
            return self.visit_binary(expr)
        if type_name == "UnaryOpValue":
            return self.visit_unary(expr)
        if type_name == "ListValue":
            return self.visit_list(expr)
        if type_name == "DictValue":
            return self.visit_dict(expr)
        if type_name == "IndexValue":
            return self.visit_index(expr)
        if type_name == "DotValue":
            return self.visit_dot(expr)
        if type_name == "FunctionCallValue":
            return self.visit_function_call(expr)
        if type_name == "SpreadValue":
            return self.visit_spread(expr)
        # Fallback to isinstance for unknown types (maintains compatibility)
        classes = _get_state_classes()
        if isinstance(expr, classes["LiteralValue"]):
            return self.visit_literal(expr)
        assert_never(expr)

    def visit_literal(self, expr: "LiteralValue") -> TExpr:
        raise NotImplementedError

    def visit_variable(self, expr: "VariableValue") -> TExpr:
        raise NotImplementedError

    def visit_action_result(self, expr: "ActionResultValue") -> TExpr:
        raise NotImplementedError

    def visit_binary(self, expr: "BinaryOpValue") -> TExpr:
        raise NotImplementedError

    def visit_unary(self, expr: "UnaryOpValue") -> TExpr:
        raise NotImplementedError

    def visit_list(self, expr: "ListValue") -> TExpr:
        raise NotImplementedError

    def visit_dict(self, expr: "DictValue") -> TExpr:
        raise NotImplementedError

    def visit_index(self, expr: "IndexValue") -> TExpr:
        raise NotImplementedError

    def visit_dot(self, expr: "DotValue") -> TExpr:
        raise NotImplementedError

    def visit_function_call(self, expr: "FunctionCallValue") -> TExpr:
        raise NotImplementedError

    def visit_spread(self, expr: "SpreadValue") -> TExpr:
        raise NotImplementedError


class ValueExprResolver(ValueExprVisitor["ValueExpr"]):
    """Resolve variables inside a ValueExpr tree without executing actions.

    Example IR:
    - y = x + 1 (where x -> LiteralValue(2))
    Produces BinaryOpValue(LiteralValue(2), +, LiteralValue(1)).
    """

    # Instance-level cache for state classes
    __slots__ = ("_resolve_variable", "_seen", "_classes")

    def __init__(
        self,
        resolve_variable: Callable[[str, set[str]], "ValueExpr"],
        seen: set[str],
    ) -> None:
        self._resolve_variable = resolve_variable
        self._seen = seen
        self._classes = _get_state_classes()

    def visit_literal(self, expr: "LiteralValue") -> "ValueExpr":
        return expr

    def visit_variable(self, expr: "VariableValue") -> "ValueExpr":
        return self._resolve_variable(expr.name, self._seen)

    def visit_action_result(self, expr: "ActionResultValue") -> "ValueExpr":
        return expr

    def visit_binary(self, expr: "BinaryOpValue") -> "ValueExpr":
        BinaryOpValue = self._classes["BinaryOpValue"]
        return BinaryOpValue(
            left=self.visit(expr.left),
            op=expr.op,
            right=self.visit(expr.right),
        )

    def visit_unary(self, expr: "UnaryOpValue") -> "ValueExpr":
        UnaryOpValue = self._classes["UnaryOpValue"]
        return UnaryOpValue(
            op=expr.op,
            operand=self.visit(expr.operand),
        )

    def visit_list(self, expr: "ListValue") -> "ValueExpr":
        ListValue = self._classes["ListValue"]
        return ListValue(elements=tuple(self.visit(item) for item in expr.elements))

    def visit_dict(self, expr: "DictValue") -> "ValueExpr":
        DictEntryValue = self._classes["DictEntryValue"]
        DictValue = self._classes["DictValue"]
        return DictValue(
            entries=tuple(
                DictEntryValue(
                    key=self.visit(entry.key),
                    value=self.visit(entry.value),
                )
                for entry in expr.entries
            )
        )

    def visit_index(self, expr: "IndexValue") -> "ValueExpr":
        IndexValue = self._classes["IndexValue"]
        return IndexValue(
            object=self.visit(expr.object),
            index=self.visit(expr.index),
        )

    def visit_dot(self, expr: "DotValue") -> "ValueExpr":
        DotValue = self._classes["DotValue"]
        return DotValue(
            object=self.visit(expr.object),
            attribute=expr.attribute,
        )

    def visit_function_call(self, expr: "FunctionCallValue") -> "ValueExpr":
        FunctionCallValue = self._classes["FunctionCallValue"]
        return FunctionCallValue(
            name=expr.name,
            args=tuple(self.visit(arg) for arg in expr.args),
            kwargs={name: self.visit(value) for name, value in expr.kwargs.items()},
            global_function=expr.global_function,
        )

    def visit_spread(self, expr: "SpreadValue") -> "ValueExpr":
        ActionCallSpec = self._classes["ActionCallSpec"]
        SpreadValue = self._classes["SpreadValue"]
        collection = self.visit(expr.collection)
        kwargs = {name: self.visit(value) for name, value in expr.action.kwargs.items()}
        action = ActionCallSpec(
            action_name=expr.action.action_name,
            module_name=expr.action.module_name,
            kwargs=kwargs,
        )
        return SpreadValue(collection=collection, loop_var=expr.loop_var, action=action)


class ValueExprSourceCollector(ValueExprVisitor[set[UUID]]):
    """Collect execution node ids that supply data to a ValueExpr tree.

    Example IR:
    - total = a + @sum(values)
    Returns the node ids that last defined `a` and the action node for sum().
    """

    def __init__(self, resolve_variable: Callable[[str], Optional[UUID]]) -> None:
        self._resolve_variable = resolve_variable

    def visit_literal(self, expr: "LiteralValue") -> set[UUID]:
        return set()

    def visit_variable(self, expr: "VariableValue") -> set[UUID]:
        source = self._resolve_variable(expr.name)
        return {source} if source is not None else set()

    def visit_action_result(self, expr: "ActionResultValue") -> set[UUID]:
        return {expr.node_id}

    def visit_binary(self, expr: "BinaryOpValue") -> set[UUID]:
        return self.visit(expr.left) | self.visit(expr.right)

    def visit_unary(self, expr: "UnaryOpValue") -> set[UUID]:
        return self.visit(expr.operand)

    def visit_list(self, expr: "ListValue") -> set[UUID]:
        sources: set[UUID] = set()
        for item in expr.elements:
            sources.update(self.visit(item))
        return sources

    def visit_dict(self, expr: "DictValue") -> set[UUID]:
        sources: set[UUID] = set()
        for entry in expr.entries:
            sources.update(self.visit(entry.key))
            sources.update(self.visit(entry.value))
        return sources

    def visit_index(self, expr: "IndexValue") -> set[UUID]:
        return self.visit(expr.object) | self.visit(expr.index)

    def visit_dot(self, expr: "DotValue") -> set[UUID]:
        return self.visit(expr.object)

    def visit_function_call(self, expr: "FunctionCallValue") -> set[UUID]:
        sources: set[UUID] = set()
        for arg in expr.args:
            sources.update(self.visit(arg))
        for arg in expr.kwargs.values():
            sources.update(self.visit(arg))
        return sources

    def visit_spread(self, expr: "SpreadValue") -> set[UUID]:
        sources = self.visit(expr.collection)
        for arg in expr.action.kwargs.values():
            sources.update(self.visit(arg))
        return sources


class ValueExprEvaluator(ValueExprVisitor[Any]):
    """Evaluate ValueExpr nodes into concrete Python values.

    Example:
    - BinaryOpValue(VariableValue("a"), +, LiteralValue(1)) becomes the
      current value of a plus 1.
    """

    def __init__(
        self,
        resolve_variable: Callable[[str], Any],
        resolve_action_result: Callable[["ActionResultValue"], Any],
        resolve_function_call: Callable[["FunctionCallValue", Sequence[Any], Mapping[str, Any]], Any],
        apply_binary: Callable[[Any, Any, Any], Any],
        apply_unary: Callable[[Any, Any], Any],
        error_factory: Callable[[str], Exception],
    ) -> None:
        self._resolve_variable = resolve_variable
        self._resolve_action_result = resolve_action_result
        self._resolve_function_call = resolve_function_call
        self._apply_binary = apply_binary
        self._apply_unary = apply_unary
        self._error_factory = error_factory

    def visit_literal(self, expr: "LiteralValue") -> Any:
        return expr.value

    def visit_variable(self, expr: "VariableValue") -> Any:
        return self._resolve_variable(expr.name)

    def visit_action_result(self, expr: "ActionResultValue") -> Any:
        return self._resolve_action_result(expr)

    def visit_binary(self, expr: "BinaryOpValue") -> Any:
        return self._apply_binary(expr.op, self.visit(expr.left), self.visit(expr.right))

    def visit_unary(self, expr: "UnaryOpValue") -> Any:
        return self._apply_unary(expr.op, self.visit(expr.operand))

    def visit_list(self, expr: "ListValue") -> Any:
        return [self.visit(item) for item in expr.elements]

    def visit_dict(self, expr: "DictValue") -> Any:
        return {self.visit(entry.key): self.visit(entry.value) for entry in expr.entries}

    def visit_index(self, expr: "IndexValue") -> Any:
        obj = self.visit(expr.object)
        idx = self.visit(expr.index)
        return obj[idx]

    def visit_dot(self, expr: "DotValue") -> Any:
        obj = self.visit(expr.object)
        if isinstance(obj, dict):
            if expr.attribute in obj:
                return obj[expr.attribute]
            raise self._error_factory(f"dict has no key '{expr.attribute}'")
        try:
            return object.__getattribute__(obj, expr.attribute)
        except AttributeError as exc:
            raise self._error_factory(f"attribute '{expr.attribute}' not found") from exc

    def visit_function_call(self, expr: "FunctionCallValue") -> Any:
        args = [self.visit(arg) for arg in expr.args]
        kwargs = {name: self.visit(value) for name, value in expr.kwargs.items()}
        return self._resolve_function_call(expr, args, kwargs)

    def visit_spread(self, expr: "SpreadValue") -> Any:
        raise self._error_factory("cannot replay unresolved spread expression")

File: src/rappel_core/runner/value_visitor.rs
//! Shared ValueExpr visitors for traversal, resolution, and evaluation.

use std::collections::{HashMap, HashSet};

use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::state::{
    ActionCallSpec, ActionResultValue, BinaryOpValue, DictEntryValue, DictValue, DotValue,
    FunctionCallValue, IndexValue, ListValue, LiteralValue, SpreadValue, UnaryOpValue,
    VariableValue,
};

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
pub enum ValueExpr {
    Literal(LiteralValue),
    Variable(VariableValue),
    ActionResult(ActionResultValue),
    BinaryOp(BinaryOpValue),
    UnaryOp(UnaryOpValue),
    List(ListValue),
    Dict(DictValue),
    Index(IndexValue),
    Dot(DotValue),
    FunctionCall(FunctionCallValue),
    Spread(SpreadValue),
}

pub struct ValueExprResolver<'a> {
    resolve_variable: &'a dyn Fn(&str, &mut HashSet<String>) -> ValueExpr,
    seen: &'a mut HashSet<String>,
}

impl<'a> ValueExprResolver<'a> {
    pub fn new(
        resolve_variable: &'a dyn Fn(&str, &mut HashSet<String>) -> ValueExpr,
        seen: &'a mut HashSet<String>,
    ) -> Self {
        Self {
            resolve_variable,
            seen,
        }
    }

    pub fn visit(&mut self, expr: &ValueExpr) -> ValueExpr {
        match expr {
            ValueExpr::Literal(value) => ValueExpr::Literal(value.clone()),
            ValueExpr::Variable(value) => (self.resolve_variable)(&value.name, self.seen),
            ValueExpr::ActionResult(value) => ValueExpr::ActionResult(value.clone()),
            ValueExpr::BinaryOp(value) => ValueExpr::BinaryOp(BinaryOpValue {
                left: Box::new(self.visit(&value.left)),
                op: value.op,
                right: Box::new(self.visit(&value.right)),
            }),
            ValueExpr::UnaryOp(value) => ValueExpr::UnaryOp(UnaryOpValue {
                op: value.op,
                operand: Box::new(self.visit(&value.operand)),
            }),
            ValueExpr::List(value) => ValueExpr::List(ListValue {
                elements: value.elements.iter().map(|item| self.visit(item)).collect(),
            }),
            ValueExpr::Dict(value) => ValueExpr::Dict(DictValue {
                entries: value
                    .entries
                    .iter()
                    .map(|entry| DictEntryValue {
                        key: self.visit(&entry.key),
                        value: self.visit(&entry.value),
                    })
                    .collect(),
            }),
            ValueExpr::Index(value) => ValueExpr::Index(IndexValue {
                object: Box::new(self.visit(&value.object)),
                index: Box::new(self.visit(&value.index)),
            }),
            ValueExpr::Dot(value) => ValueExpr::Dot(DotValue {
                object: Box::new(self.visit(&value.object)),
                attribute: value.attribute.clone(),
            }),
            ValueExpr::FunctionCall(value) => ValueExpr::FunctionCall(FunctionCallValue {
                name: value.name.clone(),
                args: value.args.iter().map(|arg| self.visit(arg)).collect(),
                kwargs: value
                    .kwargs
                    .iter()
                    .map(|(name, arg)| (name.clone(), self.visit(arg)))
                    .collect(),
                global_function: value.global_function,
            }),
            ValueExpr::Spread(value) => {
                let kwargs = value
                    .action
                    .kwargs
                    .iter()
                    .map(|(name, arg)| (name.clone(), self.visit(arg)))
                    .collect::<HashMap<_, _>>();
                let action = ActionCallSpec {
                    action_name: value.action.action_name.clone(),
                    module_name: value.action.module_name.clone(),
                    kwargs,
                };
                ValueExpr::Spread(SpreadValue {
                    collection: Box::new(self.visit(&value.collection)),
                    loop_var: value.loop_var.clone(),
                    action,
                })
            }
        }
    }
}

pub struct ValueExprSourceCollector<'a> {
    resolve_variable: &'a dyn Fn(&str) -> Option<Uuid>,
}

impl<'a> ValueExprSourceCollector<'a> {
    pub fn new(resolve_variable: &'a dyn Fn(&str) -> Option<Uuid>) -> Self {
        Self { resolve_variable }
    }

    pub fn visit(&self, expr: &ValueExpr) -> HashSet<Uuid> {
        match expr {
            ValueExpr::Literal(_) => HashSet::new(),
            ValueExpr::Variable(value) => {
                (self.resolve_variable)(&value.name).into_iter().collect()
            }
            ValueExpr::ActionResult(value) => [value.node_id].into_iter().collect(),
            ValueExpr::BinaryOp(value) => {
                let mut sources = self.visit(&value.left);
                sources.extend(self.visit(&value.right));
                sources
            }
            ValueExpr::UnaryOp(value) => self.visit(&value.operand),
            ValueExpr::List(value) => {
                let mut sources = HashSet::new();
                for item in &value.elements {
                    sources.extend(self.visit(item));
                }
                sources
            }
            ValueExpr::Dict(value) => {
                let mut sources = HashSet::new();
                for entry in &value.entries {
                    sources.extend(self.visit(&entry.key));
                    sources.extend(self.visit(&entry.value));
                }
                sources
            }
            ValueExpr::Index(value) => {
                let mut sources = self.visit(&value.object);
                sources.extend(self.visit(&value.index));
                sources
            }
            ValueExpr::Dot(value) => self.visit(&value.object),
            ValueExpr::FunctionCall(value) => {
                let mut sources = HashSet::new();
                for arg in &value.args {
                    sources.extend(self.visit(arg));
                }
                for arg in value.kwargs.values() {
                    sources.extend(self.visit(arg));
                }
                sources
            }
            ValueExpr::Spread(value) => {
                let mut sources = self.visit(&value.collection);
                for arg in value.action.kwargs.values() {
                    sources.extend(self.visit(arg));
                }
                sources
            }
        }
    }
}

pub struct ValueExprEvaluator<'a, E> {
    resolve_variable: &'a dyn Fn(&str) -> Result<serde_json::Value, E>,
    resolve_action_result: &'a dyn Fn(&ActionResultValue) -> Result<serde_json::Value, E>,
    resolve_function_call: &'a ResolveFunctionCall<'a, E>,
    apply_binary:
        &'a dyn Fn(i32, serde_json::Value, serde_json::Value) -> Result<serde_json::Value, E>,
    apply_unary: &'a dyn Fn(i32, serde_json::Value) -> Result<serde_json::Value, E>,
    error_factory: &'a dyn Fn(&str) -> E,
}

type ResolveFunctionCall<'a, E> = dyn Fn(
        &FunctionCallValue,
        Vec<serde_json::Value>,
        HashMap<String, serde_json::Value>,
    ) -> Result<serde_json::Value, E>
    + 'a;

impl<'a, E> ValueExprEvaluator<'a, E> {
    pub fn new(
        resolve_variable: &'a dyn Fn(&str) -> Result<serde_json::Value, E>,
        resolve_action_result: &'a dyn Fn(&ActionResultValue) -> Result<serde_json::Value, E>,
        resolve_function_call: &'a ResolveFunctionCall<'a, E>,
        apply_binary: &'a dyn Fn(
            i32,
            serde_json::Value,
            serde_json::Value,
        ) -> Result<serde_json::Value, E>,
        apply_unary: &'a dyn Fn(i32, serde_json::Value) -> Result<serde_json::Value, E>,
        error_factory: &'a dyn Fn(&str) -> E,
    ) -> Self {
        Self {
            resolve_variable,
            resolve_action_result,
            resolve_function_call,
            apply_binary,
            apply_unary,
            error_factory,
        }
    }

    pub fn visit(&self, expr: &ValueExpr) -> Result<serde_json::Value, E> {
        match expr {
            ValueExpr::Literal(value) => Ok(value.value.clone()),
            ValueExpr::Variable(value) => (self.resolve_variable)(&value.name),
            ValueExpr::ActionResult(value) => (self.resolve_action_result)(value),
            ValueExpr::BinaryOp(value) => {
                let left = self.visit(&value.left)?;
                let right = self.visit(&value.right)?;
                (self.apply_binary)(value.op, left, right)
            }
            ValueExpr::UnaryOp(value) => {
                let operand = self.visit(&value.operand)?;
                (self.apply_unary)(value.op, operand)
            }
            ValueExpr::List(value) => {
                let mut items = Vec::with_capacity(value.elements.len());
                for item in &value.elements {
                    items.push(self.visit(item)?);
                }
                Ok(serde_json::Value::Array(items))
            }
            ValueExpr::Dict(value) => {
                let mut map = serde_json::Map::with_capacity(value.entries.len());
                for entry in &value.entries {
                    let key_value = self.visit(&entry.key)?;
                    let key = key_value
                        .as_str()
                        .map(|value| value.to_string())
                        .unwrap_or_else(|| key_value.to_string());
                    let entry_value = self.visit(&entry.value)?;
                    map.insert(key, entry_value);
                }
                Ok(serde_json::Value::Object(map))
            }
            ValueExpr::Index(value) => {
                let object = self.visit(&value.object)?;
                let index = self.visit(&value.index)?;
                match (object, index) {
                    (serde_json::Value::Array(items), serde_json::Value::Number(idx)) => {
                        let idx = idx.as_i64().unwrap_or(-1);
                        if idx < 0 || idx as usize >= items.len() {
                            return Err((self.error_factory)("index out of range"));
                        }
                        Ok(items[idx as usize].clone())
                    }
                    (serde_json::Value::Object(map), serde_json::Value::String(key)) => map
                        .get(&key)
                        .cloned()
                        .ok_or_else(|| (self.error_factory)("dict has no key")),
                    _ => Err((self.error_factory)("unsupported index operation")),
                }
            }
            ValueExpr::Dot(value) => {
                let object = self.visit(&value.object)?;
                if let serde_json::Value::Object(map) = object {
                    return map
                        .get(&value.attribute)
                        .cloned()
                        .ok_or_else(|| (self.error_factory)("dict has no key"));
                }
                Err((self.error_factory)("attribute not found"))
            }
            ValueExpr::FunctionCall(value) => {
                let mut args = Vec::with_capacity(value.args.len());
                for arg in &value.args {
                    args.push(self.visit(arg)?);
                }
                let mut kwargs = HashMap::new();
                for (name, arg) in &value.kwargs {
                    kwargs.insert(name.clone(), self.visit(arg)?);
                }
                (self.resolve_function_call)(value, args, kwargs)
            }
            ValueExpr::Spread(_) => Err((self.error_factory)(
                "cannot replay unresolved spread expression",
            )),
        }
    }
}

pub fn resolve_value_tree(
    value: &ValueExpr,
    resolve_variable: &dyn Fn(&str, &mut HashSet<String>) -> ValueExpr,
) -> ValueExpr {
    let mut seen = HashSet::new();
    let mut resolver = ValueExprResolver::new(resolve_variable, &mut seen);
    resolver.visit(value)
}

pub fn collect_value_sources(
    value: &ValueExpr,
    resolve_variable: &dyn Fn(&str) -> Option<Uuid>,
) -> HashSet<Uuid> {
    let collector = ValueExprSourceCollector::new(resolve_variable);
    collector.visit(value)
}
